### PR TITLE
Set a specific Elixir history directory for IEx

### PR DESCRIPTION
**Pull request by josevalim**
/cc @michalmuskala

### CODE REVIEW

File: bin/iex [hunk 1]
```diff
@@ -43,6 +43,14 @@ readlink_f () {
...
+ERL="-noshell -user Elixir.IEx.CLI"
 SELF=$(readlink_f "$0")
 SCRIPT_PATH=$(dirname "$SELF")
-exec "$SCRIPT_PATH"/elixir --no-halt --erl "-noshell -user Elixir.IEx.CLI" +iex "$@"
+
+if [ -z "$HOME" ]; then
+  :
+else
+  ERL="-kernel shell_history_path \"$HOME/.elixir-history\" $ERL"
+fi
```
Comment (josevalim) [inline bin/iex]: We need someone using Windows to translate this addition to the `iex.bat` file. Help please? /cc @OnorioCatenacci @jwfearn @orderthruchaos  @robi-wan 

Is the $HOME variable even set on Windows?

**Comment by OnorioCatenacci** [comment]
It can be set. There is a variable but it is not called "HOME". I would be
happy to help but I am currently on holiday and don't have access to a
computer. I will be home after July 5.

**Comment by OnorioCatenacci** [comment]
FWIW here is a list of predefined environment variables. I think
%HomeDrive%/%HomePath% or %AppData% would work.

**Comment by OnorioCatenacci** [comment]
Sorry, neglected to paste the link in my last e-mail:

http://vlaurie.com/windows-7-tips/environment-variables-windows-vista-7.html

**Comment by robi-wan** [comment]
As @OnorioCatenacci pointed out: there is no environment variable %HOME% under Windows. %APPDATA% or% USERPROFILE% could work:
* `APPDATA=C:\Users\username\AppData\Roaming`
* `USERPROFILE=C:\Users\username`

This reminds me of [ERL-161 Consider using uniform, cross-platform way to retrieve user home directory](https://bugs.erlang.org/browse/ERL-161): Erlang internally (mostly) uses `init:get_argument(home)` to determine the $HOME folder of the user.

Would it be possible to set the variable `shell_history_path` in Elixir code (maybe somewhere in the IEx app with something like `:application.set_env(:kernel, 
 :shell_history_path, Path.join(:elixir_config.get(:home), ".elixir-history")`?

File: bin/iex [hunk 2]
```diff
@@ -43,6 +43,14 @@ readlink_f () {
...
+
+echo exec "$SCRIPT_PATH"/elixir --no-halt --erl "$ERL" +iex "$@"
```
Comment (robi-wan) [inline bin/iex]: I think this `echo` is not necessary.

**Comment by josevalim** [comment]
Unfortunately we cannot run Elixir code. Everything needs to be done with the knowledge we have in the `.bat` file. I think APPDATA is the way to go.

File: bin/iex (hunk 2 above)
Comment (josevalim) [inline bin/iex]: Good catch!

**Comment by josevalim** [comment]
Thanks @OnorioCatenacci for the help and the links. Enjoy the holidays, hopefully someone will contribute a proper version sooner. :)

**Comment by michalmuskala** [comment]
I think the erlang's `filename:basedir` (used for default location of the shell history) would by default use a path inside:
* as directed by env variable `$XDG_CACHE_HOME/erlang_history` or otherwise in `~/.cache/erlang_history` on linux
* in `~/Library/Caches/erlang_history` in macOS
* in `%LOCALAPPDATA%/erlang_history/Cache` with fallback to `%APPDATA` if `%LOCALAPPDATA%` is not set (that's apparently the case on XP) and yet another fallback to `$HOME` or `init:get_argument(home)` as the base (and that apparently can happen when running under MSYS).

To sum up, it's fairly complex :( While on unixes storing inside `~/.elixir_history` should be fine, on windows it might be harder given all those fallbacks and possibly not available variables.

I also wonder if it would make sense to set it in `elixir` itself instead of only in `iex`?

**Comment by josevalim** [comment]
> I also wonder if it would make sense to set it in elixir itself instead of only in iex?

I thought about the same but I could not come up with a good reason to do so... so I didn't. :)

**Comment by josevalim** [comment]
@michalmuskala I think we will be fine with using %APPDATA% on Windows.

File: bin/iex (hunk 1 above)
Comment (dimitarvp) [inline bin/iex]: Windows is giving you the %HOMEPATH% variable for use inside CMD. It doesn't not even contain a drive letter, for example it's `\Users\dimi` on my Win10 machine. %HOMEDRIVE% complements it, mine yields `C:`. So I imagine concatenating both gives the desired result.

@josevalim Do you need help translating only the highlighted part?

**Comment by josevalim** [comment]
Yes, only the highlighted part. I think APPDATA is probably the best choice.
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by dimitarvp** [comment]
@josevalim Are you sure about APPDATA? On my Win10 machine the `.hex`  and `.mix` directories are inside HOMEPATH and not in APPDATA.

File: bin/iex (hunk 1 above)
Comment (dimitarvp) [inline bin/iex]: @josevalim This looks to be working for me (in isolation anyway):

```
@set ERL=-noshell -user Elixir.IEx.CLI
@set DIR=%HOMEDRIVE%%HOMEPATH%
@if not exist %DIR% goto end
@set ERL=-kernel shell_history_path %DIR%\.elixir-history %ERL%
```

Or you can replace the second line with `@set DIR=%APPDATA%` if you prefer.

If you tell me which file you want this snippet in (I am guessing `iex.bat`?) I can send a PR.

Comment (josevalim) [inline bin/iex]: @dimitarvp yes, please do send a PR. Let's go with HOMEDRIVE/HOMEPATH for now for consistency then. My only question is if you have to quote the `%DIR%` variable somehow. Or at least test that the feature works if your user directory has a space on it.

Comment (dimitarvp) [inline bin/iex]: @josevalim I'll do it but can't promise it for today or Friday -- might be the weekend. I'll account for these cases in a PR:

* Directory without spaces
* Directory with spaces (you're right that quoting might be necessary)
* Directory which is an NTFS symbolic link
* Directory which is an NTFS junction point (ref: https://stackoverflow.com/questions/9042542/what-is-the-difference-between-ntfs-junction-points-and-symbolic-links)

It's definitely gonna be a bit larger than the straightforward script above, but it should be quite a bit more bulletproof.

**Comment by OnorioCatenacci** [comment]
The path may contain spaces so you should quote it just to be safe. More
likely to have issues with spaces than NTFS symbolic links or junction
points.

**Comment by dimitarvp** [comment]
@OnorioCatenacci Yep, true. I already made a slightly modified version that uses quotes and which works perfect with and without spaces in the directory name (tested it).

I'll poke around some more in the symlinks and junctions because some organizations are known to redirect people's home directories to Samba LAN mounts. It's not something you see everyday but if I can cover it within a reasonable time frame of experimentation, I will. If not, it's highly unlikely we'll be bitten by this anyway.

**Comment by OnorioCatenacci** [comment]
By the way thanks for the kind wishes José. I only look at the emails at
the end of the day to help me relax before bed. :)

**Comment by josevalim** [comment]
@dimitarvp if you have a version that works on directories with and without quotes, it is enough to get the ball rolling and for us to merge. So please do send a pull request on top of this one. :) If a fix is required for the other bits, then we can always merge it separately. Thanks!

**Comment by dimitarvp** [comment]
@josevalim Got you. I invested a few hours in the symlink / junction
functionality but it's simply too flaky to be used.

I'll send a PR during the weekend. Promise.

File: bin/iex [hunk 3]
```diff
@@ -43,6 +43,14 @@ readlink_f () {
...
 
+ERL="-noshell -user Elixir.IEx.CLI"
 SELF=$(readlink_f "$0")
 SCRIPT_PATH=$(dirname "$SELF")
-exec "$SCRIPT_PATH"/elixir --no-halt --erl "-noshell -user Elixir.IEx.CLI" +iex "$@"
+
+if [ -z "$HOME" ]; then
+  :
+else
+  ERL="-kernel shell_history_path \"$HOME/.elixir-history\" $ERL"
```
Comment (josevalim) [inline bin/iex]: Note to self: I need to verify that this also works with spaces in directory names on Unix.

**Comment by dimitarvp** [comment]
Here you go, maintainers: https://github.com/elixir-lang/elixir/pull/6254

**Comment by josevalim** [comment]
Closing this as we are having issues with double quotes on both Unix and Windows. The feature still works and I have pushed the docs directly to master. We will look into adding a new config to OTP that will make configuring this easier.

**Comment by loeschg** [comment]
In case it helps others (I had to do a little digging... wasn't sure where to find the aforementioned docs) - https://github.com/elixir-lang/elixir/commit/e6165c054805a7838c43dd44b1ea8a5fc5e49d1f
